Make the "global" word list a list by "window" and refine score

kaste opened the pull request:
1. 
 
Really having *all* words in the box is probably a mistake as it is a) a
wildly large list and b) it suggests really not available tokens from
other projects.

2.
Refine `fuzzy_score`

- Give matches at the start 2 extra points
- Do not give penalty for consecutive matching characters (!)

The higher the score the unlikely we suggest it.  So we could as well
say the `score` is a "penalty".  For consecutive matches, we still payed
"1" (as in `pos - prev`); we remove here.

This is WIP as in this commit message `consectuv<TAB>` will not suggest
"consecutive".  ST was always able to fix simple typos as well.

Other than that this algorithm mostly suggests what ST suggests on
`alt-space` in the *same* order.

kaste committed:
Make the "global" word list a list by "window"

Really having *all* words in the box is probably a mistake as is a) a
wildly large list and `b) it suggests really not available tokens from
other projects.

kaste committed:
Refine `fuzzy_score`

The diff looks larger than it maybe should.

- Give matches at the start 2 extra points
- Do not give penalty for consecutive matching characters (!)

The higher the score the unlikely we suggest it.  So we could as well
say the `score` is a "penalty".  For consecutive matches, we still payed
"1" (as in `pos - prev`); we remove here.

This is WIP as in this commit message `consectuv<TAB>` will not suggest
"consecutive".  ST was always able to fix simple typos as well.

Other than that this algorithm mostly suggests what ST suggests on
`alt-space` in the *same* order.

kaste commented:
Thanks for the project.  I think the scoring I implemented here is *almost* what ST has.  **But** ST also allows some typos to be corrected.  (Basically here in this text ST would fix "correcet" to "corrected".)  This was always useful, for example you could always write `teh<tab>` to get a `the`.

1. The history you implemented doesn't work at all for me.  ST did not had this feature, but for me the *initial* primer should be mapped to the final replacement.  (Basically what ST persists in the workspace or session files.)  For example, I write `fbs`, and then after some tabs continue to write, t.i. accepting, with `find_by_selector`.  It feels naive or natural to persist this exact mapping `fbs => find_by_selector` and prepend it to the other matching candidates. 

2. Ideally we wouldn't process the "global" list; it might grow rather fast, and as soon as the current view is large enough (as in: provides enough completion suggestions), the user might never reach it.  We often might waste the computation.

kaste committed:
Allow some typos

Especially twist like "teh" instead of "the".

kaste committed:
Cut down suggestions by scaling by 2 and dropping gt 5

kaste commented:
All WIP, 7334c77 would allow to fix some typos.  Of course, then the list of all suggestions grows.  We could maybe scale the scope, and not suggest for example anything with a scope greater than x (say 5 or 10).  For example for "el" we might suggest "replace" but we this is unlikely helpful.  It should have a score we can probably drop.

kaste commented:
Maybe if interested, just switch between the latest three commits.

kaste committed:
Do not bind for "read_only" views

Suor commented:
It's hard to review and discuss several things in the single PR. Can you split these?

kaste committed:
Reimplement "history" as a map from abbrevation to expanded word

kaste committed:
Do exclude the `initial_primer` from the word list

Usually the `initial_primer` is a best match.  Don't include it in the
word list so we don't have to skip over it initially.

Note that we always include the primer through `_matching()`. What we
don't want is: after suggesting the history item, first suggest the
primer as its probably the top match from the current view, and only
then show the next suggestion from the view.

kaste commented:
Yes, sure.  This is just a short `pu`/dev branch.

kaste commented:
Close as it is more complicated than I thought.
